runtime/jcl/common/jdk_internal_foreign_abi_UpcallStubs.cpp

```diff
@@ -65,11 +65,9 @@ Java_jdk_internal_foreign_abi_UpcallStubs_freeUpcallStub0(JNIEnv *env, jobject r
 
 	PORT_ACCESS_FROM_JAVAVM(vm);
 
-	omrthread_monitor_enter(vm->thunkHeapWrapperMutex);
+	omrthread_monitor_enter(vm->thunkHeapListMutex);
 	if (NULL != thunkAddr) {
```

Comment:
Should the dereference that happens on AIX and z/OS occur after we've acquired the mutex?

Reply:
That's unwrapping a function descriptor which are read-only.

Reply:
That space where the pointer lives may be reused, or can we assume this will be called at most once for a given (live) `thunkAddr`?

Reply:
Yes, the thunk will be automatically released from OpenJDK when the scope/session in the application is closed, in which case all resources related to the thunk are released (the space of the thunk in the heap will be reused next time and is bound to a new upcall method)

Reply:
Can you answer the "at most once" part of my question? If this might ever be called more than once for a particular address (from a single allocation), we have a problem.

Reply:
> If this might ever be called more than once for a particular address (from a single allocation), we have a problem.
Nope, it is called only once by OpenJDK when the scope/session in the application is closed.